internal/jobcontainers/oci.go

```diff
@@ -0,0 +1,77 @@
+package jobcontainers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/jobobject"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/oci"
+	"github.com/Microsoft/hcsshim/internal/processorinfo"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+// This file contains helpers for converting parts of the oci spec to useful
+// structures/limits to be applied to a job object.
+
+// Oci spec to job object limit information. Will do any conversions to job object specific values from
+// their respective OCI representations. E.g. we convert CPU count into the correct job object cpu
+// rate value internally.
+func specToLimits(ctx context.Context, s *specs.Spec) (*jobobject.JobLimits, error) {
+	// CPU limits
+	cpuNumSet := 0
+	cpuCount := uint32(oci.ParseAnnotationsCPUCount(ctx, s, oci.AnnotationContainerProcessorCount, 0))
+	if cpuCount > 0 {
+		cpuNumSet++
+	}
+
+	cpuLimit := uint32(oci.ParseAnnotationsCPULimit(ctx, s, oci.AnnotationContainerProcessorLimit, 0))
+	if cpuLimit > 0 {
+		cpuNumSet++
+	}
+
+	cpuWeight := uint32(oci.ParseAnnotationsCPUWeight(ctx, s, oci.AnnotationContainerProcessorWeight, 0))
+	if cpuWeight > 0 {
+		cpuNumSet++
+	}
+
+	if cpuNumSet > 1 {
+		return nil, fmt.Errorf("invalid spec - Windows Job Container CPU Count: '%d', Limit: '%d', and Weight: '%d' are mutually exclusive", cpuCount, cpuLimit, cpuWeight)
+	} else if cpuNumSet == 1 {
+		hostCPUCount := uint32(processorinfo.ProcessorCount())
+		if cpuCount > hostCPUCount {
+			log.G(ctx).WithFields(logrus.Fields{
+				"requested": cpuCount,
+				"assigned":  hostCPUCount,
+			}).Warn("Changing user requested CPUCount to current number of processors")
+			cpuCount = hostCPUCount
+		}
+
+		if cpuCount != 0 {
+			// Convert cpu count to cpu rate (limit). Job object API has no concept of
+			// a CPU count, so we need to convert the requested amount of processors to
+			// a cpu rate that translates to being able to use x number of processors.
+			cpuLimit = calculateJobCPURate(hostCPUCount, cpuCount)
+		} else if cpuWeight != 0 {
+			cpuWeight = calculateJobCPUWeight(cpuWeight)
+		}
```

Comment:
why not just do this above when we parse the values? We can avoid unnecessary if/else here.

Reply:
Yea I thought about this, I need to check what gets spit out if these values are 0 and get run through these.

Reply:
If you pass in 0 to calculateJobCPURate currently it will return 1 so we can't just do the below.

`cpuLimit = calculateJobCPURate(hostCPUCount, uint32(oci.ParseAnnotationsCPULimit(ctx, s, oci.AnnotationContainerProcessorLimit, 0)))`

I know we've being using HCS' logic here for the utility function but would it make more sense to return 0 here instead of 1 if you pass in 0 for the second parameter? When we actually call `jobobject.SetResourceLimits` later on, it won't set anything if it's 0 which is what we'd want if the client didn't actually pass anything in, but passing in 1 is definitely not what we want, it would set the lowest possible cpu rate.